### DISCUSSION TITLE
Is there an inverse of `basis.interpolate()`?

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#816] [Q&A] [answered]
Just that... Is there an inverse of `basis.interpolate()?`

**Comment by kinnala** [comment]
In theory it is L2 projection, i.e. skfem.utils.projection. However, I'm still not completely happy about skfem.utils.projection. I'm working on a new version which will hopefully also be useful for this.

This is what I have in mind, proj() is the new function, and has similar syntax to asm():
```python
basis = CellBasis(mesh, ElementTriP1())

# w is like the last parameter to forms
y1 = proj(lambda w: w.x[0], basis)

# take derivative with gradient averaging
x = mesh.p[0]
y2 = proj(lambda w: w['u'].grad[0], basis, u=x)

# real derivative
y3 = proj(lambda w: w['u'].grad[0], basis.with_element(ElementTriP0()), u=basis.interpolate(x))

# inverse
y4 = proj(lambda w: w['u'], basis, u=basis.interpolate(x))
# y4 should be almost equal to x
```

**Reply by gatling-nrl** [reply]
So the breaking change, making you want to rename the function to `proj` is going from `projection(lambda x: ...)` to `proj(lambda w: ...)` and using `w.x` to get the `x` that you had previously.

I guess I'm opposed to this change. Let me first wave away my pedantic objections. I don't like nouns as function names. `projection()` and `refined()` really bother me. Functions are _verbs_.  `.project()` and `.refine()` are so much more natural. I hesitate to critique, but names are powerful things and I don't think you should take them lightly. You already have a deprecated `project()` and now you're tinkering with deprecating `projection()` in favor of `proj`. This will cause painful confusion to new users, you can be sure of that.

I think the concept of "projecting into the finite element space" is one of the most central concepts. I tried to weight it heavily in my extended docs draft proposal. It is the main mechanism for taking things you understand... like math functions, python functions, experimental data... into the world of the finite element solver.

You have several ways of doing that currently.... I think `get_dofs()` and this `utils.proj*` function we're talking about. But I vote to promote it well beyond afterthought. I think it is one of the most important features of skfem and might deserve a serious discussion about the right computer-sciencey abstraction.  _Espeically_ if you're considering adding another name to the skfem namespace.

**Reply by kinnala** [reply]
refine() in 1.0.0 was something that had a side effect and modified the existing object. We wanted to get rid of those because of the implications to objects depending on those, e.g., CellBasis, and called the new methods refined() etc. so it makes clear it returns a copy. Now refine() does not exist anymore but I don't think there is a need to change the names again.

**Reply by kinnala** [reply]
Let me add that the name proj above is somewhat arbitrary, I just wanted to demonstrate the functionality, it might be possible to implement in projection in a backwards compatible manner.

**Reply by gatling-nrl** [reply]
> but I don't think there is a need to change the names again

I strongly agree.  I also understand better your thoughts: `refine()` to modify and existing object in place and `refined()` to return a _new_ object and leave the original reference alone.

Frequently I am left wondering which of those two things is happening when I call numpy functions. (Sadly, even familiar ones...) So I often have to resort to google, once again.  Nevertheless, I prefer verbs and clear docstrings... and very much prefer backward compatibility over consistent or pedantic naming.

**Reply by gatling-nrl** [reply]
> Let me add that the name proj above is somewhat arbitrary, I just wanted to demonstrate the functionality, it might be possible to implement in projection in a backwards compatible manner.

Can your new ideas not be implemented in `AbstractBasis.proj(ect)`?  It is just such an important (I think) feature of skfem that it seems unintuitve to have it relegated to something called `utils` in the first place. Isn't it a first class type operation in the semantics?  Can you even do the projecting without the `AbstractBasis` object?

**Reply by kinnala** [reply]
Here is a simple draft implementation for `proj` above:
```python
def proj(fun, basis, **kwargs):
    from skfem.models import mass
    M = mass.assemble(basis)
    f = LinearForm(lambda v, w: fun(w) * v).assemble(basis, **kwargs)
    return solve(M, f)
```
As you can see, it is much simpler than the implementation of `projection` and takes care of almost all special cases supported by `projection`. Adding this to `skfem.utils` is technically not a backwards incompatible change (in the spirit of semantic versioning) since we are adding another function. Only removing or changing functionality is backwards incompatible.

**Reply by kinnala** [reply]
> > Let me add that the name proj above is somewhat arbitrary, I just wanted to demonstrate the functionality, it might be possible to implement in projection in a backwards compatible manner.
> 
> Can your new ideas not be implemented in `AbstractBasis.proj(ect)`? It is just such an important (I think) feature of skfem that it seems unintuitve to have it relegated to something called `utils` in the first place. Isn't it a first class type operation in the semantics? Can you even do the projecting without the `AbstractBasis` object?

Implementation inside `AbstractBasis` would be something like:
```python
class AbstractBasis:
    ...
    def project(self, fun, **kwargs):
        from skfem.models import mass
        from skfem.utils import solve
        from skfem.assembly import LinearForm
        M = mass.assemble(basis)
        f = LinearForm(lambda v, w: fun(w) * v).assemble(self, **kwargs)
        return solve(M, f)
```

**Reply by kinnala** [reply]
This is how the snippets would look like:
```python
basis = CellBasis(mesh, ElementTriP1())

# w is like the last parameter to forms
y1 = basis.project(lambda w: w.x[0])

# take derivative with gradient averaging
x = mesh.p[0]
y2 = basis.project(lambda w: w['u'].grad[0], u=x)

# real derivative
basis0 = basis.with_element(ElementTriP0())
y3 = basis0.project(lambda w: w['u'].grad[0], u=basis.interpolate(x))

# inverse
y4 = basis.project(lambda w: w['u'], u=basis.interpolate(x))
# y4 should be almost equal to x
```
I think it kind of makes sense.

**Comment by kinnala** [comment, marked as answer]
Alright, to answer the original question, with the existing `projection` it would like this:
```python
import numpy as np
from skfem import *
m = MeshTri()
basis = Basis(m, ElementTriP1())
x1 = m.p[0]  # the function f(x) = x
x1_interp = basis.interpolate(x1)
x2 = projection(LinearForm(lambda v, w: x1_interp * v), basis)
print(np.linalg.norm(x1 - x2))
```
which will print `2.6766507790745728e-16`.

**Reply by kinnala** [reply]
I guess `projection` can already technically do anything that my other suggestion can, I'd just somehow enjoy being able to pass `x1_interp` via a keyword argument to `w` and not by referencing it directly inside the function definition.

**Reply by gatling-nrl** [reply]
> `x1 = m.p[0]  # the function f(x) = x`

Only works because you're using `P1`, right? Generically you could use `x1 = skfem.projection(lambda x: x, basis)`?

**Reply by gatling-nrl** [reply]
> which will print 2.6766507790745728e-16.

`2.6208083778984373e-16`

Why did we get slightly different answers? I copy pasted your code.  Is that a CPU architecture thing? A dependency version thing? It bothers me.

**Reply by gatling-nrl** [reply]
```python
m = skfem.MeshTri()
basis = skfem.Basis(m, skfem.ElementTriP1())
x1 = skfem.projection(lambda x: x[0], basis)  # previously x1 = m.p[0] # the function f(x) = x
x1_interp = basis.interpolate(x1)
x2 = skfem.projection(skfem.LinearForm(lambda v, w: x1_interp * v), basis)
print(np.linalg.norm(x1 - x2))
```
`4.562732988774191e-16`

**Reply by kinnala** [reply]
My guess is that because inside `projection` there is `solve` and those sparse matrix solves are known to be pretty sensitive to small things at the floating point level. What kind of CPU you are using? This is mine:
```
[tom@tunkki:~]$ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Tavujärjestys:                   Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          8
On-line CPU(s) list:             0-7
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Malli:                           142
Mallinimi:                       Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
```
Might also depend on some of the compilation settings of the sparse solvers.

**Reply by gatling-nrl** [reply]
```
$ wmic cpu get caption, deviceid, name, numberofcores, maxclockspeed, status
Caption                                 DeviceID  MaxClockSpeed  Name                                      NumberOfCores  Status
Intel64 Family 6 Model 158 Stepping 12  CPU0      3600           Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz  8              OK
```

**Reply by kinnala** [reply]
This shows I'm using OpenBLAS and not MKL:
```
In [2]: numpy.show_config()
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
blas_opt_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
lapack_opt_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/5gq47dyyzjz4p2h603drb2srs66hr2fr-lapack-3/lib', '/nix/store/2cz8569qf7khb1cqdli88y1ywwyqw5b3-blas-3/lib']
```
Moreover, 
```
In [3]: numpy.__version__
Out[3]: '1.20.2'

In [4]: import scipy

In [5]: scipy.__version__
Out[5]: '1.6.1'
```

**Reply by gatling-nrl** [reply]
```
np.show_config()
blas_info:
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\lib']
    include_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\include']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\lib']
    include_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\include']
    language = f77
lapack_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas']
    library_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\lib']
    language = f77
lapack_opt_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas', 'cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\lib']
    language = f77
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/ggatling/Miniconda3\\Library\\include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL
```
```
np.__version__
'1.21.2'
scipy.__version__
'1.7.1'
```

**Reply by kinnala** [reply]
It seems that SuperLU is included in scipy GitHub repository so basically the sparse solver version depends on the scipy version.

**Reply by kinnala** [reply]
Alright, I know that the version of numpy in conda is optimized for performance and will most likely use some specific compilation flags or different BLAS version. I'm using a build script from nixpkgs which is just a really basic build: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/scipy/default.nix

Also the version number difference could explain it.
